src/appleseed/renderer/kernel/lighting/lighttree.cpp

```diff
@@ -212,36 +230,47 @@ float LightTree::recursive_node_update(size_t node_index, size_t node_level)
         const auto& child1 = m_nodes[node_index].get_child_node_index();
         const auto& child2 = m_nodes[node_index].get_child_node_index() + 1;
 
-        float luminance1 = recursive_node_update(child1, node_level + 1);
-        float luminance2 = recursive_node_update(child2, node_level + 1);
+        float luminance1 = recursive_node_update(node_index, child1, node_level + 1);
+        float luminance2 = recursive_node_update(node_index, child2, node_level + 1);
 
         luminance = luminance1 + luminance2;
     }
     else
     {
         // Access the light intensity value.
         size_t item_index = m_nodes[node_index].get_item_index();
-        size_t light_source_index = m_items[item_index].m_light_index;
-        Spectrum spectrum = m_light_sources[light_source_index]->get_intensity();
-
-        luminance = foundation::average_value(spectrum);
+        size_t light_source_index = m_items[item_index].m_light_source_index;
+        luminance = m_light_sources[light_source_index]->get_intensity();
 
         // Modify the tree depth if the branch is deeper than the other branches
         // visited so far.
         if (m_tree_depth < node_level)
             m_tree_depth = node_level;
+
+        // If the light is emitting triangle, store the info about it's position in the light
+        // tree.
+        if (m_light_sources[light_source_index]->get_type() == LightSource::EMT)
+        {
+            m_light_sources[light_source_index]->set_tree_index(node_index);
+        }
     }
 
     m_nodes[node_index].set_luminance(luminance);
     m_nodes[node_index].set_level(node_level);
+    if (node_index != 0)
+        m_nodes[node_index].set_parent(parent_index);
+    else
+        m_nodes[node_index].set_root();
 
     return luminance;
 }
 
-std::pair<size_t, float> LightTree::sample(
+std::tuple<int, size_t, float> LightTree::sample(
     const foundation::Vector3d&     surface_point,
     float                           s) const
 {
+    assert (is_built());
```

Comment:
Remove space after `assert`, it's not a keyword or an operator but a function (a macro really but still).